File: puzzle2.c
#include <limits.h>
#include <math.h>
#include <stdio.h>

struct child {
    int gender, weight;
};

int find_minimum(struct child c[], int length) {
    int result = INT_MAX;

    for (int i = 0; i < length; i++) {
        if (c[i].gender == 0)
            for (int j = 0; j < length; j++)
                if (c[j].gender == 1)
                    result = abs(c[i].weight - c[j].weight) < result ? abs(c[i].weight - c[j].weight) : result;
    }

    return result;
}

int main(void) {
    struct child c[] = {
        {0, 1},
        {0, 13},
        {1, 9},
        {0, 7},
        {1, 15}};

    printf("Output: %d \n", find_minimum(c, sizeof(c) / sizeof(c[0])));
    return 0;
}

File: puzzle2.ts

interface Child {
    gender: number;
    weight: number;
}

// Big O Notation: O(2n)
const findMinimum = (children: Child[]): number => {
    let result = 10_000;

    for (let i = 0; i < children.length; i++) {
        if (children[i].gender == 0) {
            for (let j = 0; j < children.length; j++) {
                if (children[j].gender == 1)
                    result = Math.abs(children[i].weight - children[j].weight) < result ? Math.abs(children[i].weight - children[j].weight) : result;
            }
        }
    }

    return result;
};

// Test Case 1
// const totalChildren = 5;
// const genders = [0, 0, 1, 0, 1];
// const weights = [1, 13, 9, 7, 15];
// Output : 2

// Test Case 2
// const totalChildren = 3;
// const genders = [0, 0, 1];
// const weights = [1, 2, 2];
// Output : 0

const children: Child[] = [
    {
        gender: 0,
        weight: 1
    },
    {
        gender: 0,
        weight: 13
    },
    {
        gender: 1,
        weight: 9
    },
    {
        gender: 0,
        weight: 7
    },
    {
        gender: 1,
        weight: 15
    }
];

const result = findMinimum(children);
console.log(result);
